src/controllers/hid/hidcontroller.cpp

```diff
@@ -194,22 +194,52 @@ QByteArray HidController::getInputReport(unsigned int reportID) {
 
 bool HidController::poll() {
     Trace hidRead("HidController poll");
-
-    // This loop risks becoming a high priority endless loop in case processing
-    // the mapping JS code takes longer than the controller polling rate.
-    // This could stall other low priority tasks.
-    // There is no safety net for this because it has not been demonstrated to be
-    // a problem in practice.
+    // This function reads the available HID Input Reports using hidapi.
+    // Important to know is, that this reading is not a hardware operation,
+    // instead it reads previously received HID Input Reports from a ring buffer.
+    // Depending on the hidapi backend implementation the ring buffer is either part
+    // of the hidapi implementation or the OS kernel.
+    // The size of the ring buffer also depends on the hidapi backend:
+    // - hidraw(2048 bytes) - Linux Kernel API
+    // - libusb(30 reports) - BSD (alternative Linux userspace implementation)
+    // - mac(30 reports)
+    // - windows(64 reports)
+    // Behavior:
+    // - If the polling rate is to slow, multiple HID Input Reports will be processed at the same time
+    // - If the mapping processes the Input Reports too slow, whole Input Reports will be omitted if the ring buffer is full
+    // - If the processing of all received HID Input Reports in the JS mapping code takes longer than the interval between incoming reports,
+    //   it could stall other low priority tasks.
+
+    constexpr int kSmallestRingBufferInReports = 30;
+    constexpr int kSmallestRingBufferInBytes = 2048;
+
+    int inputReportCounter = 0;
+    int inputReportByteCounter = 0;
     while (true) {
         int bytesRead = hid_read(m_pHidDevice, m_pPollData[m_pollingBufferIndex], kBufferSize);
         if (bytesRead < 0) {
             // -1 is the only error value according to hidapi documentation.
+            qCWarning(m_logOutput) << "Unable to read HID Input Reports from" << getName() << ":"
+                                   << mixxx::convertWCStringToQString(
+                                              hid_error(m_pHidDevice),
+                                              kMaxHidErrorMessageSize);
             DEBUG_ASSERT(bytesRead == -1);
             return false;
         } else if (bytesRead == 0) {
-            // No packet was available to be read
+            // No packets left to be read
+            if ((inputReportCounter >= kSmallestRingBufferInReports) ||
+                    (inputReportByteCounter >= kSmallestRingBufferInBytes)) {
```

Comment:
I'm still struggling to understand what this actually detects.
From what I can tell, it basically counts how many reports / bytes we've read and if we exceeded the buffer given by us from hidapi, we assume we dropped packets?

Reply:
If those really are the limits of the hidapi internal buffer, does this actually ever happen? How can we poll packets more often than there are packets in the buffer?

Reply:
From my understanding right now, the only way the warning case can even trigger is in a false-positive case where the limits are wrong.
Moreover, I'm also questioning how useful this information actually is when we allow false positives at all:
> If we are at or above the limits, **it can be**, that reports are dropped. -> Print warning

Reply:
The limit depend on the hidapi backen and also the implementation, as I described in the comment on top. The compare is greater or equal.
It depends on the implementation of the backend, if a read from the ring buffer can only read once or is constantly filled with the nex item if you read the next.

Reply:
In any case there is a performance problem, because the system is on the edge. So the warning definitally indicates a real problem.

Reply:
I still don't understand it to be honest, but thats likely because I neither know hidapi or the hidcontroller code and making assumptions of how they work.